### PR TITLE
Supercharge pyvi...

### PR DESCRIPTION
**Pull request by dexterous**
... to leave python behind once it's job is done!

### CODE REVIEW

**Review by anandology — CHANGES_REQUESTED** [review]
Thanks @dexterous. importlib is much better way to find the source. 

I think it would be better to use environment variable for setting the editor rather than using flags. For example, how would you use emacs? @nibrahim has a version of this script that works with emacs. using env vars will make it truely open. 

```
$ export PYVI_COMMAND=less
$ pyvi os.path
```

You may have to go with os.system instead of exec because the command may have flags like `vim -R` or `emacs -nw` etc. Also, please keep the default command to `vim -R` as vim is most commonly available command and I'm not sure if `view` is as common in all the distros.

Also, can you please make it into two functions, `find_source` and `view_file`? 

I would like to print the source path so that I know where the file is after coming out of it. The print was added deliberately, not for debugging.

**Comment by anandology** [comment]
OMG, this PR was pending from 2018? Sorry @dexterous, I completely missed it. Thanks for following up with more changes.

**Comment by dexterous** [comment]
Well, let me preface what I'm going to say with the fact that this is just a script I keep coming back to when I want to fiddle with Python. I don't actually use the script itself.

Lots to unpack from your reply, mostly because this has been open for so long. 🙂 

* I didn't consider alternate editors because the script was literally called `pyvi`; but if we want to consider alternative editors, I'd figure the `EDITOR` env var would be more suitable as most distros use it indicate the default editor anyway (with the script defaulting to `vim`, or should that be `vi`?). In any case that opens up the can of worms that requires us to figure out the different options to put each specific editor in readonly mode. Not to mention, invoking one of the hardest problems in CS, (re)naming the script!
* The `-g` option seemed useful because you may not always have a display; in which case it provides a quick option to switch over to gVim when one is available. I was toying with an `--editor` option to the script, but shelved that for now. I guess, we could use `EDITOR` in combination with other env vars to make this determination. That said, the only env var I can think of right now is `DISPLAY` and I don't know if that can be used reliably in our case. Not to mention the fact that this would further compound the complexity of the above issue of maintaining/looking-up command line options for the GUI versions of the editor. At the least we'd be "duplicating" them in some form for GUI/CLI versions of the editor.
* `view` is available wherever `vim` is; in fact the `vim` package installer installs `view` simply as a symlink to `vim` and it essentially simulates `vim -R`. See my [which-view](https://repl.it/@dexterous/which-view?v=1) repl.it instance for ref. Also, I always add the `-M` option onto `view` to make the buffer unmodifiable as an extra measure.
* We can definitely keep the line that prints the file source in; I just figured it was redundant considering the editor displays is anyway.
* As for splitting `open_module` into `find_source` and `view_file`, that was my first impulse as well. But I quickly figured that needed more finesse. I'd much prefer to wrap the essence of `open_module` into a class called `ModuleSource` that takes the module name as a ctor param. It would then have methods to ensure that the module has a source; IIRC C exts don't exhibit a source file, and failed lookups in `find_spec` just return `None`. See [importlib-fu](https://repl.it/@dexterous/importlib-fu?v=1). But that's a project for another day.

Anyway, what I really meant to say is that I use `pyvi` at all; it just gave me an excuse to tinker with the Python import mechanism and `exec*` (which, BTW, we should stick to for this tool as there's no point having something as light as this keep the parent Python process hanging around once it's handed the file off to the editor). I just keep coming back to it to play with a part of Python that I don't usually get to when building web apps. 😉 This PR was something I created way back when... I can't even remember. I think only the first 3 commits are from the original PR and Github's been appending my subsequent commits to it ever since.

So, basically, take it as you will. I'll keep coming back to this to work on the stuff we deliberated on above (and would be happy to keep deliberating), but feel free to adopt the changes from this without waiting for me to finish the stuff.

It was great conversing with you again @anandology. It's been a while since we bumped heads at the Python community meetups in B'lore or on the Bangpypers list.
Cheers.
-- @dexterous

**Comment by dexterous** [comment]
@anandology,
I'm putting down a list of things I think we can do to close this out:
- [X] ~Add `--editor` option~
- [X] ~Maintain lookup table of editor options to ensure readonly mode~
- [X] ~Check `EDITOR` env var to determine application to hand off to~
- [X] ~If `EDITOR` not set, default to `vim`, if not available use `vi`, if not available use `nano`~
- [x] Use `vim -RM` as default command to open editor
- [x] Substitute editor command with contents of `PYVI_EDITOR` if non-empty
- [x] Refactor `open_module` into `ModuleSource`
- [x] Handle case where `find_spec` returns `None`
- [x] Handle case where `ModuleSpec` has no `origin`
- [x] Handle case where `ModuleSpec.origin` is binary (`.so`/`.pyc`)
- [x] Handle case where `ModuleSpec.origin` is `'built-in'`

-- @dexterous

**Comment by nibrahim** [comment]
This whole business of `--editor` then `envars` then defaults is ideally served by a ChainMap.[this](https://github.com/anandology/hacks/pull/1#issuecomment-788994155
) is an example.  I'm not sure why I'm receiving emails from this issue but nevertheless.

**Comment by dexterous** [comment]
@nibrahim,
Whoa! Long time no see, man. Good to hear from you, even if it is from a rouge Github comment routing. 😄 

I think the link in your comment is broken, but yeah, a Chainmap was going to be my solution for this as well.

That said, I've been unimpressed by most implementations of Chainmap, esp as they were applied to config lookup for CLI tools and I was going to use this as a test bed to build my own.

Nice talking to you again man.
-- @dexterous

**Comment by anandology** [comment]
On Tue, Mar 2, 2021 at 9:01 PM Saager Mhatre <notifications@github.com>
wrote:

> @anandology <https://github.com/anandology>,
> I'm putting down a list of things I think we can do to close this out:
>
>    - Refactor open_module into ModuleSource
>
> ok.

>
>    - Add --editor option
>    - Maintain lookup table of editor options to ensure readonly mode
>
> No. We are not doing these two.

>
>    - Check EDITOR env var to determine application to hand off to
>    - If EDITOR not set, default to vim, if not available use vi, if not
>    available use nano
>
> No. If we are going to EDITOR when we need to figure out how to set
readonly option for every editor. We are not doing this. The command is
going to be `vim -R` unless `PYVI_EDITOR` is set.

>
>    - Handle case where find_spec returns None
>    - Handle case where ModuleSpec has no origin
>    - Handle case where ModuleSpec.origin is binary (.so/.pyc)
>    - Handle case where ModuleSpec.origin is "built-in"
>
> ok.

**Comment by dexterous** [comment]
@anandology,
Updated the list above.

-- @dexterous

**Comment by dexterous** [comment]
Reworked
```bash
$ for mod in builtins math non_existent site; do echo -e "./pyvi ${mod}\n"; ./pyvi "${mod}"; echo -e "$?\n====="; done
./pyvi builtins

'builtins' is a built-in
1
=====
./pyvi math

'math' is a binary (.so)
Searched in:
	/Users/smhatre/repos/mine/hacks
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python39.zip
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload
	/usr/local/lib/python3.9/site-packages
1
=====
./pyvi non_existent

No module named 'non_existent'
Searched in:
	/Users/smhatre/repos/mine/hacks
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python39.zip
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9
	/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload
	/usr/local/lib/python3.9/site-packages
1
=====
./pyvi site

/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site.py
0
=====
$
```

Also:
```bash
$ PYVI_EDITOR='gvim -R -M' ./pyvi site
/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site.py
$ echo $?
0
```

Let me know what you think.
-- d

**Comment by anandology** [comment]
Thanks @dexterous. Merged!
